Hypothesis ≥ 6.130.13 on Python 3.9 doesn't support the LL(1) parser anymore

[CPython 3.9 can optionally still use the older LL(1) parser (Python 3.8 and lower), instead of the newer LL(1) parser (Python 3.10 and higher).](https://docs.python.org/3/whatsnew/3.9.html#new-parser) The LL(1) parser does not understand the full set of syntax legalized with Python 3.9… in particular, it cannot parse `with (expr1 as expr2, expr3 as expr4)` statements correctly.

As of hypothesis-python-6.130.13, the code no longer parses under the LL(1) parser. For example, using `pytest` and `coverage` in a Python 3.9 `hatch` environment with `PYTHONOLDPARSER=1` set yields the following syntax error [non-public bits have been redacted]:

~~~~
ImportError while loading conftest '.../tests/conftest.py'.
tests/__init__.py:26: in <module>
    import hypothesis
.../.local/share/hatch/env/virtual/.../lib/python3.9/site-packages/hypothesis/__init__.py:29: in <module>
    from hypothesis.core import example, find, given, reproduce_failure, seed
.../.local/share/hatch/pythons/3.9/python/lib/python3.9/ast.py:50: in parse
    return compile(source, filename, mode, flags,
E     File ".../.local/share/hatch/env/virtual/.../lib/python3.9/site-packages/hypothesis/core.py", line 1108
E       BuildContext(data, is_final=is_final) as context,
E                                             ^
E   SyntaxError: invalid syntax
.../.local/share/hatch/env/virtual/.../lib/python3.9/site-packages/coverage/inorout.py:509: CoverageWarning: Module ... was never imported. (module-not-imported)
  self.warn(f"Module {pkg} was never imported.", slug="module-not-imported")
~~~~

(The same happens in a PyPy 3.9 environment with `PYTHONOLDPARSER=1`.)

---

The issue is due to f151eec1e28b07f78837c08806401cd292ff6949 (6.130.13), which specifically refactored this snippet of code into the Py3.9/LL(1)-incompatible form, starting from the compatible form. I suggest reverting this snippet to the LL(1)-compatible form.

--- a/AUTHORS.rst
+++ b/AUTHORS.rst
@@ -120,6 +120,7 @@ their individual contributions.
 * `Louis Taylor <https://github.com/kragniz>`_
 * `Luke Barone-Adesi <https://github.com/baluke>`_
 * `Lundy Bernard <https://github.com/lundybernard>`_
+* `Marco Ricci <https://github.com/the-13th-letter>`_
 * `Marco Sirabella <https://www.github.com/mjsir911>`_
 * `marekventur <https://www.github.com/marekventur>`_
 * `Marius Gedminas <https://www.github.com/mgedmin>`_ (marius@gedmin.as)

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,7 @@
+RELEASE_TYPE: patch
+
+This patch restores compatibility when using `the legacy Python 3.9 LL(1)
+parser <https://docs.python.org/3/whatsnew/3.9.html#new-parser>`__, which
+was accidentally broken since :ref:`version 6.130.13 <v6.130.13>`.
+
+Thanks to Marco Ricci for this fix!

--- a/hypothesis-python/src/hypothesis/core.py
+++ b/hypothesis-python/src/hypothesis/core.py
@@ -1102,18 +1102,19 @@ class StateForActualGivenExecution:
 
         # self.test_runner can include the execute_example method, or setup/teardown
         # _example, so it's important to get the PRNG and build context in place first.
-        with (
-            local_settings(self.settings),
-            deterministic_PRNG(),
-            BuildContext(data, is_final=is_final) as context,
-        ):
-            # providers may throw in per_case_context_fn, and we'd like
-            # `result` to still be set in these cases.
-            result = None
-            with data.provider.per_test_case_context_manager():
-                # Run the test function once, via the executor hook.
-                # In most cases this will delegate straight to `run(data)`.
-                result = self.test_runner(data, run)
+        #
+        # NOTE: For compatibility with Python 3.9's LL(1) parser, this is written as
+        # three nested with-statements, instead of one compound statement.
+        with local_settings(self.settings):
+            with deterministic_PRNG():
+                with BuildContext(data, is_final=is_final) as context:
+                    # providers may throw in per_case_context_fn, and we'd like
+                    # `result` to still be set in these cases.
+                    result = None
+                    with data.provider.per_test_case_context_manager():
+                        # Run the test function once, via the executor hook.
+                        # In most cases this will delegate straight to `run(data)`.
+                        result = self.test_runner(data, run)
 
         # If a failure was expected, it should have been raised already, so
         # instead raise an appropriate diagnostic error.

